src/GafferUIModule/PathListingWidgetBinding.cpp

```diff
@@ -1025,6 +1038,61 @@ class PathModel : public QAbstractItemModel
 			return false;
 		}
 
+		void updateHeaderData()
+		{
+			if( m_headerDataState != State::Dirty )
+			{
+				return;
+			}
+
+			std::vector<CellVariants> newHeaderData;
+			for( auto &column : m_columns )
+			{
+				newHeaderData.push_back( column->headerData() );
+			}
+
+			if( m_headerData == newHeaderData )
+			{
+				m_headerDataState = State::Clean;
+				return;
+			}
+
+			queueEdit(
+
+				[this, newHeaderData = std::move( newHeaderData )] () mutable {
+
+					m_headerData.swap( newHeaderData );
+					m_headerDataState = State::Clean;
+					headerDataChanged( Qt::Horizontal, 0, m_headerData.size() - 1 );
+
+				}
+
+			);
+		}
+
+		// State transitions :
+		//
+		// - Unrequested->Dirty : When first queried.
+		// - Dirty->Clean : When updated.
+		// - Clean->Dirty : When path changes.
+		enum class State
+		{
+			// Initial state. Not yet requested by clients
+			// of the model, therefore not yet computed, and not
+			// in need of consideration during recursive updates.
+			Unrequested,
+			// Computed and up to date.
+			Clean,
+			// Stale data that needs recomputing.
+			Dirty
+		};
+
+		static bool dirtyIfUnrequested( std::atomic<State> &state ) // MAYBE LEAVE THIS WEHERE IT WAS?
```

Comment:
I don't think I have much of an opinion on where this lives ... it's not externally visible. On the other hand, this isn't new, but I find the name dirtyIfUnrequested kinda unwieldy ... it could just be named "request"? To me it is pretty easy to understand that requesting something if it's already clean is a no-op, but this name feels confusing.

Reply:
I'd completely forgotten about my big all-caps comment - I had intended to remove it (when I wrote it, I didn't think I was going to be using `dirtyIfUnrequested()`. I think I see your point about the name, but I've left it for now, as I'm more worried about getting the new behaviour right than changing the old naming.